test-harness/tests/test_cache.py

```diff
@@ -0,0 +1,476 @@
+                SearchCriteria(1, True),
+            'Coud not retrieve Mesos state cache': SearchCriteria(1, True),
+        }
+
+        # Break marathon
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='always_bork')
+
+        ar = nginx_class()
+        url = ar.make_url_from_path(
+            '/agent/de1baf83-c36c-4d23-9cb0-f89f596cd6ab-S1/agent-1/blah/blah')
+
+        with GuardedAR(ar):
+            lbf = LineBufferFilter(filter_regexp,
+                                   timeout=(FIRST_POLL_PERIOD + 1),
+                                   line_buffer=ar.stderr_line_buffer)
+
+            resp = requests.get(url,
+                                allow_redirects=False,
+                                headers=valid_user_header)
+            lbf.scan_log_buffer()
+
+        assert resp.status_code == 503
+        assert lbf.extra_matches == {}
+
+    def test_if_broken_mesos_does_not_break_marathon_cache(
+            self, nginx_class, mocker, valid_user_header):
+        filter_regexp = {
+            'Mesos state request failed: invalid response status: 500':
+                SearchCriteria(1, True),
+            'Marathon apps cache has been successfully updated': SearchCriteria(1, True),
+        }
+
+        # Break marathon
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='always_bork')
+
+        mocker.send_command(endpoint_id='http://127.0.0.1:8080',
+                            func_name='enable_nginx_task')
+
+        ar = nginx_class()
+        url = ar.make_url_from_path('/service/nginx/bar/baz')
+
+        with GuardedAR(ar):
+            lbf = LineBufferFilter(filter_regexp,
+                                   timeout=(FIRST_POLL_PERIOD + 1),
+                                   line_buffer=ar.stderr_line_buffer)
+
+            resp = requests.get(url,
+                                allow_redirects=False,
+                                headers=valid_user_header)
+            lbf.scan_log_buffer()
+
+        assert resp.status_code == 200
+        req_data = resp.json()
+        assert req_data['endpoint_id'] == 'http://127.0.0.1:16001'
+
+        assert lbf.extra_matches == {}
+
+    def test_if_changing_marathon_apps_is_reflected_in_cache(
+            self, nginx_class, valid_user_header, mocker):
+        poll_period = 4
+        ar = nginx_class(poll_period=poll_period, cache_expiry=3)
+        url = ar.make_url_from_path('/service/nginx/bar/baz')
+
+        with GuardedAR(ar):
+            resp = requests.get(url,
+                                allow_redirects=False,
+                                headers=valid_user_header)
+            assert resp.status_code == 404
+
+            mocker.send_command(endpoint_id='http://127.0.0.1:8080',
+                                func_name='enable_nginx_task')
+
+            time.sleep(poll_period)
```

Comment:
It does not matter, but not for the reason you mentioned - the refresh *has* to happen before we execute mocker.send_command() as we want to observe the the registering new marathon apps data.

Timeline is as follows:
* nginx starts
* cache updates are scheduled:
  * CACHE_FIRST_POLL_DELAY_SECONDS by default is 2
  * CACHE_POLL_PERIOD_SECONDS is overriden to 4
  * CACHE_EXPIRATION_SECONDS is overriden to 3
* first request is made. It forces cache refresh that sets `marathonleader_last_refresh` if it happens before first poll, it uses the cache data if it happens after the first poll. This way or another we are guaranteed to have some initial data in the cache before we execute the command and we are 100% certain what this data is.
* we change the mesos state data.
* poll_period > cache_expiry, so we are guaranteed to have the cache expired by the time we perform next query. It will be refreshed by either a timer refresh or request-enforced refresh.
* we inspect response data to see if the mesos state change was reflected in the cache.

Hope it helps.